File: src/bin/0084-largest-rectangle-in-histogram.py
# @leetup=info id=84 lang=python3 slug=largest-rectangle-in-histogram




from typing import Optional

# @leetup=code
class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        # rectable:
        # area = current_bar_height * (rightmost_bar_index - leftmost_bar_index + 1)
        # all(height[i] >= current_bar_height for i in range(current_index...(rightmost_bar_index + 1)))

        # Monotonic stack (Stores index). Increasing.
        # height[i] < height[i + 1]]

        # Sentinel
        heights = [0] + heights + [0]

        stack = []
        res = 0
        for i, n in enumerate(heights):
            while stack and heights[stack[-1]] > n:
                prev_i = stack.pop()
                h = heights[prev_i]
                if stack:
                    w = i - stack[-1] - 1
                    area = h * w
                    res = max(res, area)
            stack.append(i)
        return res


# @leetup=code

File: src/bin/0084-largest-rectangle-in-histogram.rs
// @leetup=info id=84 lang=rust slug=largest-rectangle-in-histogram


// Ref: https://leetcode.com/problems/largest-rectangle-in-histogram/solutions/28917/ac-python-clean-solution-using-stack-76ms/
// @leetup=code
impl Solution {
    pub fn largest_rectangle_area(heights: Vec<i32>) -> i32 {
        let heights = {
            let mut h = vec![0];
            h.extend(heights.into_iter());
            h.push(0);
            h
        };
        let mut stack = Vec::new();
        let mut res = 0;
        for (i, &n) in heights.iter().enumerate() {
            while let Some(true) = stack.last().map(|i| heights[*i] > n) {
                let prev_i = stack.pop().unwrap();
                let h = heights[prev_i];
                if let Some(&left_index) = stack.last() {
                    //      |-----|
                    //      |     |--|
                    //   |--|     |  |
                    //   |  |     |  |
                    //            ^ i
                    //          ^ prev_i
                    //   ^ left_index

                    // for i in (left_index + 1..prev_i + 1) {
                    //    assert_eq!(heigt[i] >= height[prev_i])
                    //
                    let width = i - (left_index + 1);

                    let area = h * width as i32;
                    res = res.max(area);
                }
            }
            stack.push(i);
        }
        res

    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::largest_rectangle_area;
}
